README.md
# yui-layout
yui-layout源码  

使用 yui-layout ，能够使你不再聚焦于css，而向编写js配置项，从而实现任意场景布局，特别是在大屏自适应兼容开发场景中。  

如果你正在开发一个**大屏**项目，而你对页面**自适应布局**又没有什么即时有效的解决方法，或许yui-layout就能完美的帮你解决你当前遇到的困难，并且这个布局过程，或许只需要几分钟时间。 

##### 源码位置
```
cd src/packages/layout
```
##### 打包发布
```
yarn run package

cd yui-layout

npm version patch
or
npm verion minor
or
npm version major

npm publish
```

##### 版本管理
npm的发包需要遵循语义化版本，一个版本号包含三个部分：MAJOR.MINOR.PATCH  
- MAJOR 表示主版本号，当你做了不兼容的API修改
- MINOR 表示次版本号，当你做了向下兼容的功能性新增；
- PATCH 表示修订号,当你做了向下兼容的问题修正;  

我们可以使用npm version 命令来自动修改版本号，比如：
```
// version = v1.0.0

npm version patch // v1.0.1
npm version prepatch // v1.0.2-0
npm version minor // v1.1.0
npm version major // v2.0.0
```
一般来说还有先行版本，测试版本等，他们这样命名
- 3.1.0-alpha.0
- 3.1.0-beta.0  

如果我们发布先行版本，`npm version prepatch` 命令得出的版本号好像就不够规范了，我们只能使用 `npm version 1.0.0-alpha.1` 指定版本号，不过还好，在 npm 6.4.0 之后，我们可以使用 --preid 参数：

```
npm version prerelease --preid=alpha
```

CHANGELOG.md


# 0.2.0 (2023-04-16)


### Features

* 初始化yui-layout源码 ([d636ae4](https://github.com/Y-Cwei/yui-layout/commit/d636ae48bebed7aa319ccd013109474b7e8db02e))

src/pages/home/index.js
import header from '@/pages/home/header'
import cardOne from '@/pages/home/card-one'
import cardTwo from '@/pages/home/card-two'
import cardThree from '@/pages/home/card-three'
import cardFour from '@/pages/home/card-four'
import cardFive from '@/pages/home/card-five'
import cardSix from '@/pages/home/card-six'
import cardMiddle from '@/pages/home/card-middle'

// 设计稿尺寸
const screenWidth = 1920 // 设计图宽度
const screenHeight = 1080 // 设计图高度
const cardWidth = 470 // 卡片宽度
const cardHeight = 310 // 卡片高度
const cardBgColor = 'rgb(29, 47, 86, 0.3)'
const appBgColor = '#161c2c'
const margin = 20

// 计算各区域尺寸
const headerHeight = 80
const cardOneTop = headerHeight + margin
const cardTwoTop = cardOneTop
const cardThreeTop = cardTwoTop + cardHeight + margin
const cardFiveTop = cardThreeTop + cardHeight + margin
const cardSixTop = cardFiveTop
const cardMiddleTop = cardFiveTop
const cardMiddleWidth = screenWidth - cardWidth * 2 - margin * 4

// 高度百分比
function getHeightPercent (height) {
  return getPercent(height, screenHeight)
}
// 宽度百分比
function getWidthPercent (width) {
  return getPercent(width, screenWidth)
}
function getPercent (a, b) {
  return (a / b * 100) + '%'
}

export default {
  name: 'home', // 布局名称
  widthRange: [1300, 1920], // 宽度范围，从 widthRange[0] ~ 100%
  heightRange: [700, 1080], // 高度范围，从 heightRange[0] ~ 100%
  style: { // 定义样式
    width: '100%',
    height: '100%',
    background: appBgColor
  },
  components: [ // 组件配置
    {
      name: 'homeHeader', // 组件唯一标识
      style: {
        top: 0,
        left: 0,
        right: 0,
        height: getHeightPercent(headerHeight) // 为了自适应，宽高使用百分比
      },
      render: header // vue组件
    },
    {
      name: 'cardOne',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardOneTop),
        left: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardOne
    },
    {
      name: 'cardTwo',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardTwoTop),
        right: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardTwo
    },
    {
      name: 'cardThree',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardThreeTop),
        left: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardThree
    },
    {
      name: 'cardFour',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardThreeTop),
        right: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardFour
    },
    {
      name: 'cardFive',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardFiveTop),
        left: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardFive
    },
    {
      name: 'cardSix',
      style: {
        width: getWidthPercent(cardWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardSixTop),
        right: getWidthPercent(margin),
        background: cardBgColor
      },
      render: cardSix
    },
    {
      name: 'cardMiddle',
      style: {
        width: getWidthPercent(cardMiddleWidth),
        height: getHeightPercent(cardHeight),
        top: getHeightPercent(cardMiddleTop),
        left: getWidthPercent(cardWidth + 2 * margin),
        background: cardBgColor
      },
      render: cardMiddle
    }
  ]
}


yui-layout/yui-layout.common.js
/******/ (function() { // webpackBootstrap
/******/ 	"use strict";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/define property getters */
/******/ 	!function() {
/******/ 		// define getter functions for harmony exports
/******/ 		__webpack_require__.d = function(exports, definition) {
/******/ 			for(var key in definition) {
/******/ 				if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 					Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 				}
/******/ 			}
/******/ 		};
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/hasOwnProperty shorthand */
/******/ 	!function() {
/******/ 		__webpack_require__.o = function(obj, prop) { return Object.prototype.hasOwnProperty.call(obj, prop); }
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	!function() {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = function(exports) {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/publicPath */
/******/ 	!function() {
/******/ 		__webpack_require__.p = "";
/******/ 	}();
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// ESM COMPAT FLAG
__webpack_require__.r(__webpack_exports__);

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  "default": function() { return /* binding */ entry_lib; }
});

;// CONCATENATED MODULE: ./node_modules/@vue/cli-service/lib/commands/build/setPublicPath.js
/* eslint-disable no-var */
// This file is imported into lib/wc client bundles.

if (typeof window !== 'undefined') {
  var currentScript = window.document.currentScript
  if (false) { var getCurrentScript; }

  var src = currentScript && currentScript.src.match(/(.+\/)[^/]+\.js(\?.*)?$/)
  if (src) {
    __webpack_require__.p = src[1] // eslint-disable-line
  }
}

// Indicate to webpack that this file can be concatenated
/* harmony default export */ var setPublicPath = (null);

;// CONCATENATED MODULE: ./node_modules/thread-loader/dist/cjs.js!./node_modules/babel-loader/lib/index.js??clonedRuleSet-40.use[1]!./node_modules/@vue/vue-loader-v15/lib/loaders/templateLoader.js??ruleSet[1].rules[3]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=template&id=296b1056&scoped=true&
var render = function render() {
  var _vm = this,
      _c = _vm._self._c;

  return _c('div', {
    staticClass: "yui-container"
  }, _vm._l(_vm.layouts, function (layout, index) {
    return _c('div', {
      key: index,
      style: _vm.getContainerStyle(layout),
      attrs: {
        "id": layout.name,
        "refs": layout.name
      }
    }, _vm._l(layout.components, function (comp, key) {
      return _c('section', {
        key: key,
        style: comp.style,
        attrs: {
          "id": comp.name
        }
      }, [_c(comp.render, {
        tag: "component",
        attrs: {
          "refs": comp.name
        }
      })], 1);
    }), 0);
  }), 0);
};

var staticRenderFns = [];

;// CONCATENATED MODULE: ./node_modules/thread-loader/dist/cjs.js!./node_modules/babel-loader/lib/index.js??clonedRuleSet-40.use[1]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=script&lang=js&
/* harmony default export */ var layoutvue_type_script_lang_js_ = ({
  name: 'YuiLayout',
  props: {
    layouts: {
      type: Array,
      required: true
    }
  },
  methods: {
    getContainerStyle(layout) {
      return { ...layout.style,
        minWidth: layout.widthRange[0] + 'px',
        minHeight: layout.heightRange[0] + 'px'
      };
    }

  }
});
;// CONCATENATED MODULE: ./src/packages/layout/index.vue?vue&type=script&lang=js&
 /* harmony default export */ var packages_layoutvue_type_script_lang_js_ = (layoutvue_type_script_lang_js_); 
;// CONCATENATED MODULE: ./node_modules/mini-css-extract-plugin/dist/loader.js??clonedRuleSet-12.use[0]!./node_modules/css-loader/dist/cjs.js??clonedRuleSet-12.use[1]!./node_modules/@vue/vue-loader-v15/lib/loaders/stylePostLoader.js!./node_modules/postcss-loader/dist/cjs.js??clonedRuleSet-12.use[2]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=style&index=0&id=296b1056&prod&scoped=true&lang=css&
// extracted by mini-css-extract-plugin

;// CONCATENATED MODULE: ./src/packages/layout/index.vue?vue&type=style&index=0&id=296b1056&prod&scoped=true&lang=css&

;// CONCATENATED MODULE: ./node_modules/@vue/vue-loader-v15/lib/runtime/componentNormalizer.js
/* globals __VUE_SSR_CONTEXT__ */

// IMPORTANT: Do NOT use ES2015 features in this file (except for modules).
// This module is a runtime utility for cleaner component module output and will
// be included in the final webpack user bundle.

function normalizeComponent(
  scriptExports,
  render,
  staticRenderFns,
  functionalTemplate,
  injectStyles,
  scopeId,
  moduleIdentifier /* server only */,
  shadowMode /* vue-cli only */
) {
  // Vue.extend constructor export interop
  var options =
    typeof scriptExports === 'function' ? scriptExports.options : scriptExports

  // render functions
  if (render) {
    options.render = render
    options.staticRenderFns = staticRenderFns
    options._compiled = true
  }

  // functional template
  if (functionalTemplate) {
    options.functional = true
  }

  // scopedId
  if (scopeId) {
    options._scopeId = 'data-v-' + scopeId
  }

  var hook
  if (moduleIdentifier) {
    // server build
    hook = function (context) {
      // 2.3 injection
      context =
        context || // cached call
        (this.$vnode && this.$vnode.ssrContext) || // stateful
        (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext) // functional
      // 2.2 with runInNewContext: true
      if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
        context = __VUE_SSR_CONTEXT__
      }
      // inject component styles
      if (injectStyles) {
        injectStyles.call(this, context)
      }
      // register component module identifier for async chunk inferrence
      if (context && context._registeredComponents) {
        context._registeredComponents.add(moduleIdentifier)
      }
    }
    // used by ssr in case component is cached and beforeCreate
    // never gets called
    options._ssrRegister = hook
  } else if (injectStyles) {
    hook = shadowMode
      ? function () {
          injectStyles.call(
            this,
            (options.functional ? this.parent : this).$root.$options.shadowRoot
          )
        }
      : injectStyles
  }

  if (hook) {
    if (options.functional) {
      // for template-only hot-reload because in that case the render fn doesn't
      // go through the normalizer
      options._injectStyles = hook
      // register for functional component in vue file
      var originalRender = options.render
      options.render = function renderWithStyleInjection(h, context) {
        hook.call(context)
        return originalRender(h, context)
      }
    } else {
      // inject component registration as beforeCreate hook
      var existing = options.beforeCreate
      options.beforeCreate = existing ? [].concat(existing, hook) : [hook]
    }
  }

  return {
    exports: scriptExports,
    options: options
  }
}

;// CONCATENATED MODULE: ./src/packages/layout/index.vue



;


/* normalize component */

var component = normalizeComponent(
  packages_layoutvue_type_script_lang_js_,
  render,
  staticRenderFns,
  false,
  null,
  "296b1056",
  null
  
)

/* harmony default export */ var layout = (component.exports);
;// CONCATENATED MODULE: ./src/packages/index.js

const components = [layout];

const install = Vue => {
  components.forEach(component => {
    Vue.component(component.name, component);
  });
};

/* harmony default export */ var src_packages = (install);
;// CONCATENATED MODULE: ./node_modules/@vue/cli-service/lib/commands/build/entry-lib.js


/* harmony default export */ var entry_lib = (src_packages);


module.exports = __webpack_exports__;
/******/ })()
;

yui-layout/yui-layout.umd.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["yui-layout"] = factory();
	else
		root["yui-layout"] = factory();
})((typeof self !== 'undefined' ? self : this), function() {
return /******/ (function() { // webpackBootstrap
/******/ 	"use strict";
/******/ 	// The require scope
/******/ 	var __webpack_require__ = {};
/******/ 	
/************************************************************************/
/******/ 	/* webpack/runtime/define property getters */
/******/ 	!function() {
/******/ 		// define getter functions for harmony exports
/******/ 		__webpack_require__.d = function(exports, definition) {
/******/ 			for(var key in definition) {
/******/ 				if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
/******/ 					Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
/******/ 				}
/******/ 			}
/******/ 		};
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/hasOwnProperty shorthand */
/******/ 	!function() {
/******/ 		__webpack_require__.o = function(obj, prop) { return Object.prototype.hasOwnProperty.call(obj, prop); }
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/make namespace object */
/******/ 	!function() {
/******/ 		// define __esModule on exports
/******/ 		__webpack_require__.r = function(exports) {
/******/ 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 			}
/******/ 			Object.defineProperty(exports, '__esModule', { value: true });
/******/ 		};
/******/ 	}();
/******/ 	
/******/ 	/* webpack/runtime/publicPath */
/******/ 	!function() {
/******/ 		__webpack_require__.p = "";
/******/ 	}();
/******/ 	
/************************************************************************/
var __webpack_exports__ = {};
// ESM COMPAT FLAG
__webpack_require__.r(__webpack_exports__);

// EXPORTS
__webpack_require__.d(__webpack_exports__, {
  "default": function() { return /* binding */ entry_lib; }
});

;// CONCATENATED MODULE: ./node_modules/@vue/cli-service/lib/commands/build/setPublicPath.js
/* eslint-disable no-var */
// This file is imported into lib/wc client bundles.

if (typeof window !== 'undefined') {
  var currentScript = window.document.currentScript
  if (false) { var getCurrentScript; }

  var src = currentScript && currentScript.src.match(/(.+\/)[^/]+\.js(\?.*)?$/)
  if (src) {
    __webpack_require__.p = src[1] // eslint-disable-line
  }
}

// Indicate to webpack that this file can be concatenated
/* harmony default export */ var setPublicPath = (null);

;// CONCATENATED MODULE: ./node_modules/thread-loader/dist/cjs.js!./node_modules/babel-loader/lib/index.js??clonedRuleSet-82.use[1]!./node_modules/@vue/vue-loader-v15/lib/loaders/templateLoader.js??ruleSet[1].rules[3]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=template&id=296b1056&scoped=true&
var render = function render() {
  var _vm = this,
      _c = _vm._self._c;

  return _c('div', {
    staticClass: "yui-container"
  }, _vm._l(_vm.layouts, function (layout, index) {
    return _c('div', {
      key: index,
      style: _vm.getContainerStyle(layout),
      attrs: {
        "id": layout.name,
        "refs": layout.name
      }
    }, _vm._l(layout.components, function (comp, key) {
      return _c('section', {
        key: key,
        style: comp.style,
        attrs: {
          "id": comp.name
        }
      }, [_c(comp.render, {
        tag: "component",
        attrs: {
          "refs": comp.name
        }
      })], 1);
    }), 0);
  }), 0);
};

var staticRenderFns = [];

;// CONCATENATED MODULE: ./node_modules/thread-loader/dist/cjs.js!./node_modules/babel-loader/lib/index.js??clonedRuleSet-82.use[1]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=script&lang=js&
/* harmony default export */ var layoutvue_type_script_lang_js_ = ({
  name: 'YuiLayout',
  props: {
    layouts: {
      type: Array,
      required: true
    }
  },
  methods: {
    getContainerStyle(layout) {
      return { ...layout.style,
        minWidth: layout.widthRange[0] + 'px',
        minHeight: layout.heightRange[0] + 'px'
      };
    }

  }
});
;// CONCATENATED MODULE: ./src/packages/layout/index.vue?vue&type=script&lang=js&
 /* harmony default export */ var packages_layoutvue_type_script_lang_js_ = (layoutvue_type_script_lang_js_); 
;// CONCATENATED MODULE: ./node_modules/mini-css-extract-plugin/dist/loader.js??clonedRuleSet-54.use[0]!./node_modules/css-loader/dist/cjs.js??clonedRuleSet-54.use[1]!./node_modules/@vue/vue-loader-v15/lib/loaders/stylePostLoader.js!./node_modules/postcss-loader/dist/cjs.js??clonedRuleSet-54.use[2]!./node_modules/@vue/vue-loader-v15/lib/index.js??vue-loader-options!./src/packages/layout/index.vue?vue&type=style&index=0&id=296b1056&prod&scoped=true&lang=css&
// extracted by mini-css-extract-plugin

;// CONCATENATED MODULE: ./src/packages/layout/index.vue?vue&type=style&index=0&id=296b1056&prod&scoped=true&lang=css&

;// CONCATENATED MODULE: ./node_modules/@vue/vue-loader-v15/lib/runtime/componentNormalizer.js
/* globals __VUE_SSR_CONTEXT__ */

// IMPORTANT: Do NOT use ES2015 features in this file (except for modules).
// This module is a runtime utility for cleaner component module output and will
// be included in the final webpack user bundle.

function normalizeComponent(
  scriptExports,
  render,
  staticRenderFns,
  functionalTemplate,
  injectStyles,
  scopeId,
  moduleIdentifier /* server only */,
  shadowMode /* vue-cli only */
) {
  // Vue.extend constructor export interop
  var options =
    typeof scriptExports === 'function' ? scriptExports.options : scriptExports

  // render functions
  if (render) {
    options.render = render
    options.staticRenderFns = staticRenderFns
    options._compiled = true
  }

  // functional template
  if (functionalTemplate) {
    options.functional = true
  }

  // scopedId
  if (scopeId) {
    options._scopeId = 'data-v-' + scopeId
  }

  var hook
  if (moduleIdentifier) {
    // server build
    hook = function (context) {
      // 2.3 injection
      context =
        context || // cached call
        (this.$vnode && this.$vnode.ssrContext) || // stateful
        (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext) // functional
      // 2.2 with runInNewContext: true
      if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
        context = __VUE_SSR_CONTEXT__
      }
      // inject component styles
      if (injectStyles) {
        injectStyles.call(this, context)
      }
      // register component module identifier for async chunk inferrence
      if (context && context._registeredComponents) {
        context._registeredComponents.add(moduleIdentifier)
      }
    }
    // used by ssr in case component is cached and beforeCreate
    // never gets called
    options._ssrRegister = hook
  } else if (injectStyles) {
    hook = shadowMode
      ? function () {
          injectStyles.call(
            this,
            (options.functional ? this.parent : this).$root.$options.shadowRoot
          )
        }
      : injectStyles
  }

  if (hook) {
    if (options.functional) {
      // for template-only hot-reload because in that case the render fn doesn't
      // go through the normalizer
      options._injectStyles = hook
      // register for functional component in vue file
      var originalRender = options.render
      options.render = function renderWithStyleInjection(h, context) {
        hook.call(context)
        return originalRender(h, context)
      }
    } else {
      // inject component registration as beforeCreate hook
      var existing = options.beforeCreate
      options.beforeCreate = existing ? [].concat(existing, hook) : [hook]
    }
  }

  return {
    exports: scriptExports,
    options: options
  }
}

;// CONCATENATED MODULE: ./src/packages/layout/index.vue



;


/* normalize component */

var component = normalizeComponent(
  packages_layoutvue_type_script_lang_js_,
  render,
  staticRenderFns,
  false,
  null,
  "296b1056",
  null
  
)

/* harmony default export */ var layout = (component.exports);
;// CONCATENATED MODULE: ./src/packages/index.js

const components = [layout];

const install = Vue => {
  components.forEach(component => {
    Vue.component(component.name, component);
  });
};

/* harmony default export */ var src_packages = (install);
;// CONCATENATED MODULE: ./node_modules/@vue/cli-service/lib/commands/build/entry-lib.js


/* harmony default export */ var entry_lib = (src_packages);


/******/ 	return __webpack_exports__;
/******/ })()
;
});